Utilities/Mergers/include/Mergers/Merger.h

```diff
@@ -0,0 +1,73 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef ALICEO2_MERGERS_H
+#define ALICEO2_MERGERS_H
+
+/// \file Merger.h
+/// \brief Definition of O2 Merger, v1.0
+///
+/// \author Piotr Konopka, piotr.jan.konopka@cern.ch
+
+#include "Mergers/MergerConfig.h"
+#include "Mergers/MergerCache.h"
+#include "Mergers/MergeInterface.h"
+
+#include <Framework/Task.h>
+
+#include <TObject.h>
+
+#include <memory>
+
+namespace o2
+{
+namespace mergers
+{
+
+/// \brief Merger data processor class.
+///
+/// Mergers are DPL devices able to merge ROOT objects produced in parallel.
+class Merger : public framework::Task
+{
+ public:
+  /// \brief Default constructor. It expects merger configuration and subSpec of output channel.
+  Merger(MergerConfig, header::DataHeader::SubSpecificationType);
+  /// \brief Default destructor.
+  ~Merger();
+
+  /// \brief Merger init callback.
+  void init(framework::InitContext& ctx) override;
+  /// \brief Merger process callback.
+  void run(framework::ProcessingContext& ctx) override;
+
+ private:
+  // todo: maybe the pointer vector to the extracted objects should be stored all the time
+  std::function<void()> prepareTimerCallback(framework::InitContext& ictx) const;
+  std::vector<TObject*> unpackObjects(TObject* obj);
+  void mergeCache();
+  void publish(framework::DataAllocator& allocator);
+
+  void cleanCacheAfterMerging();
+  void cleanCacheAfterPublishing();
+
+  bool shouldPublish(framework::ProcessingContext&);
+  bool shouldMergeCache(framework::ProcessingContext& ctx);
+
+ private:
+  header::DataHeader::SubSpecificationType mSubSpec;
+  MergerCache mCache;
+  std::unique_ptr<TObject> mMergedObjects;
```

Comment:
so you have 1 merger per 1 root object - how do you handle the more realistic case of multiple objects coming from an analysis/calibration device? How does it handle the case of missing inputs? (the makeup of the output deos not have to be the same on every push from the producer...)

Reply:
I expect that one ROOT object can actually contain several objects, this is what I use unpackObjects() for. I plan to provide generic possibility to unpack TCollections and using a custom unpacking function is already possible.
https://github.com/AliceO2Group/AliceO2/blob/f91c6d84b29795a29ff0be55330c547c9bde6697/Utilities/Mergers/src/Merger.cxx#L243
I did not indeed explicitly cover the possibility of missing objects inside collections and having different order inside. It could still be done by implementing custom unpack() and merge() functions, but I can think of how to do it in the library itself.

Reply:
so you go through the unpack cycle every time you want to merge in something?

Reply:
missing objects - or rather possibly different output on every push from a producer - is needed if you want to be able to handle things asynchronously (which you do). One cannot rely on ordering of objects in incoming data. My suggestion is to merge apples with apples based on some identifier - with ROOT one can use e.g. object name (as the online HLT mergers did). In O2 we might want to generalize the id/name to the level of O2 metadata (e.g. store the name in the header stack, there is a provision to do this) - so dispatch can happen before deserialization so that work can be e.g. dispatched to separate threads/devices or whathaveyou.

Reply:
or e.g. put a hash of the name in subspecification

Reply:
> so you go through the unpack cycle every time you want to merge in something?

Yes. I thought of keeping the unpacked pointers to the stored merged object as a performance improvement, but I do not see how can I not unpack each new object to merge in.

> My suggestion is to merge apples with apples based on some identifier - with ROOT one can use e.g. object name (as the online HLT mergers did). 

OK, agreed.

Reply:
> Yes. I thought of keeping the unpacked pointers to the stored merged object as a performance improvement, but I do not see how can I not unpack each new object to merge in.

well, I think you have to unpack incoming objects if they are TCollections, we just have to keep in mind that this in essence is just a work around to avoid ROOT slow TMethodCall stuff. Once unpacked it is hard to keep track of exactly what kind of containers they were in, so for simplicity's sake let's not try to reconstruct the original exact structure :) - so the merged objects can be unpacked (lifetime managed by something smart) and referencable by the id (like a map<string,TObject> or so).
We do have to keep track of the logical structure (say if we had nested containers etc. which has been the case in run2) - the simple solution in the HLT was to change the name of the unpacked object to reflect it's place in the "container tree". Not sure this is the most elegant solution but it was simple and it worked. Here we could place this path-like information in the metadata without renaming the actual root objects. I'm open to suggestions.
For user defined types (MergeInterface) - a merge() will be implemented so there should be no need to fall back to unpacking for performance reasons - that could be a config option whether you want to force this or not.
Or - we could agree to disallow the direct use of TCollections and only allow compliant containers, something we would have to cook up. Then we don't need to unpack, less code to worry about, less bugs to hunt related to the ownership non-semantics of TCollection.